### PR TITLE
feat: add `no-signal-mutable-updates` ESLint rule to enforce immutabi…

### PR DESCRIPTION
**Pull request by eneajaho**
…lity of signals

### CODE REVIEW

**Comment by JrPribs** [comment]
So this fails when a signal variable is attempted to be reassigned instead of set/updated? Do people do that?

**Comment by SchroederSteffen** [comment]
I support the intention, but isn't this somewhat covered by [prefer-signals](https://github.com/angular-eslint/angular-eslint/blob/main/packages/eslint-plugin/docs/rules/prefer-signals.md) because it expects you to set signal properties to `readonly` preventing such reassignment?

Screenshot 2026-02-09 at 08 41 31

**Comment by reduckted** [comment]
> I support the intention, but isn't this somewhat covered by [prefer-signals](https://github.com/angular-eslint/angular-eslint/blob/main/packages/eslint-plugin/docs/rules/prefer-signals.md) because it expects you to set signal properties to readonly preventing such reassignment?

That rule prevents you from changing the _signal_ instance. This new rule appears to prevent mutation of the value that the signal provides.

This rule doesn't appear to handle indirect mutation.
```ts
let s = signal([1,2,3]);
s.push(4); // This would be reported as a problem.

let array = s();
array.push(4); // It doesn't look like this would be reported as a problem.
```

**Review by JamesHenry — CHANGES_REQUESTED** [review]
Please can you address the feedback from @reduckted

**Comment by reduckted** [comment]
```ts
let items: ModelSignal<number[]>;
mutate(items());
function mutate(array: number[]) {
    array.push(1);
}
```

I think a better approach would be to look at the type of `T` in the `WritableSignal<T>` or `ModelSignal<T>`. If that type is mutable, then an error should be reported.

**Comment by JamesHenry** [comment]
@eneajaho Just following up on this one please

**Review by reduckted — CHANGES_REQUESTED** [review]
Have I misunderstood what this rule is trying to achieve?

My interpretation was that it should prevent you from storing a mutable value in a signal. Not that it should allow mutable values but prevent you from mutating them. 🤔

File: packages/eslint-plugin/tests/rules/no-signal-mutable-updates/cases.ts [hunk 1]
```diff
@@ -0,0 +1,611 @@
...
+    userData.set({ name: 'newName' });
+  `,
+  // Proper WritableSignal usage with .update()
+  `
+    let userData: WritableSignal<{ name: string }>;
+    userData.update(current => ({ ...current, name: 'newName' }));
+  `,
+  // Proper array signal update with .set()
+  `
+    let items: WritableSignal<number[]>;
```
[the remark this answers is not in the document]
Comment (eneajaho) [reply packages/eslint-plugin/tests/rules/no-signal-mutable-updates/cases.ts:37]: The idea is to have this rule that we can drop in existing projects and it should show warnings or errors based on what they have currently, without having to go and update their types to make them immutable, and sometimes devs they don't own the types (types coming from api generators for example).

[the remark this answers is not in the document]
Comment (reduckted) [reply packages/eslint-plugin/tests/rules/no-signal-mutable-updates/cases.ts:37]: Mapped types like [`DeepReadonly`](https://github.com/ts-essentials/ts-essentials/tree/master/lib/deep-readonly) can be used to make mutable types immutable.

I like the idea of this rule, but unfortunately I can still find ways to get past it and mutate the signal value.
```ts
let userData: Signal<{ a: number }[]>;
let x = userData();
x[0].a = 42;
```
```ts
let userData: Signal<{ a: number }[]>;
mutate(userData());

function mutate(array: { a: number }[]) {
   array[0].a = 42;
}
```
```ts
let userData: Signal<{ a: number }>;
mutate(userData());

function mutate(x: { a: number }) {
    x.a = 42;
}
```

I've created a PR with an alternative approach to this, where the rule will check that the type of value in the signal is immutable. #2989. It still needs work, but I thought I'd post it and see whether the complexity of determining whether a type is immutable is worth it.

**Comment by eneajaho** [comment]
How should we continue with this?

**Comment by eneajaho** [comment]
Hi @JamesHenry another ping here 🙌

File: packages/eslint-plugin/tests/rules/no-signal-mutable-updates/cases.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (JamesHenry) [reply packages/eslint-plugin/tests/rules/no-signal-mutable-updates/cases.ts:37]: @eneajaho I'm sorry this has stalled for a while. I'd love to know your thoughts on @reduckted's alternative approached proposed here.

In general I would say part of the reason this has stalled is because it seems like it's going to have gaps, and the gaps could lead to confusing experiences. I really like the intention of helping protect people against writing bad signal code, but my instinct was also that it was going to be type based...

What are your concerns, if any, with the approach in #2989?

[the remark this answers is not in the document]
Comment (eneajaho) [reply packages/eslint-plugin/tests/rules/no-signal-mutable-updates/cases.ts:37]: Hi @JamesHenry , no worries at all — I appreciate you circling back!

I've looked at @reduckted's approach in #2989 and I think it's a genuinely interesting angle, but my concern is about the adoption path it requires.

The type-based approach enforces that signal values are declared with immutable types (e.g. readonly properties). That's a sound correctness model, but it means users have to actively update their type definitions before the rule provides any value — and in many real-world projects, that's not always possible. Types frequently come from third-party libraries, API code generators, or shared packages that the consuming team doesn't control. Asking those teams to mark everything readonly as a prerequisite creates a significant barrier to adoption.

The rule in this MR takes a different stance: it detects mutations at the call site — the .set(), .update(), and .mutate() patterns where mutable operations actually happen. This means a team can drop it into an existing project and immediately get signal warnings without touching any type definitions. It meets projects where they are, rather than where we'd ideally want them to be.

That said, I don't think these two rules are in conflict — they're actually complementary. Teams that do control their types and want stricter, type-level guarantees could enable #2989's rule alongside this one. Both can live in a project simultaneously, each catching a different class of issues. Rather than choosing one over the other, having both available gives teams the flexibility to adopt as much strictness as their codebase allows.

I do understand the concern about gaps — detecting mutation through object references is inherently harder than a type-level check, and there will be cases it can't catch. But I'd argue that catching the common cases immediately, with zero migration cost, is more valuable for the broader ecosystem than a theoretically complete solution that requires upfront type changes most teams can't easily make.

Happy to discuss further or add more test coverage for the edge cases you have in mind. Let me know what would help move this forward.

**Comment by JamesHenry** [comment]
@eneajaho I'm sorry that I didn't help you get this in before 22, as you can imagine people were clamoring for that.

If we can please rebase this and make sure all the edge cases raised in comments are captured, then we can press ahead with it 👍
